internal/validation/ingress/ingress.go

```diff
@@ -0,0 +1,81 @@
+package ingress
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/kong/go-kong/kong"
+	"github.com/sirupsen/logrus"
+	netv1 "k8s.io/api/networking/v1"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/failures"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/versions"
+	kongv1alpha1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1alpha1"
+)
+
+type routeValidator interface {
+	Validate(context.Context, *kong.Route) (bool, string, error)
+}
+
+func ValidateIngress(
+	ctx context.Context,
+	routesValidator routeValidator,
+	parserFeatures parser.FeatureFlags,
+	kongVersion semver.Version,
+	ingress *netv1.Ingress,
+) (bool, string, error) {
+	discardLogger := logrus.New()
+	discardLogger.Out = io.Discard
+	failuresCollector, err := failures.NewResourceFailuresCollector(discardLogger)
+	if err != nil {
+		return false, "", err
+	}
+	var icp kongv1alpha1.IngressClassParametersSpec
+	if kongVersion.LT(versions.ExplicitRegexPathVersionCutoff) {
+		icp.EnableLegacyRegexDetection = true
+	}
+	result := parser.IngressesV1ToKongServices(
+		parserFeatures,
+		[]*netv1.Ingress{ingress},
+		icp,
+		&parser.ObjectsCollector{}, // It's irrelevant for validation.
+		failuresCollector,
+	)
+
+	var errMsgs []string
+	for _, f := range failuresCollector.PopResourceFailures() {
+		errMsgs = append(errMsgs, f.Message())
+	}
+	if len(errMsgs) > 0 {
+		return false, validationMsg(errMsgs), nil
```

Comment:
I think that in the case of `len(errMsgs) > 0` that we got from `failuresCollector` we could have a validation message prefix other than `Ingress failed schema validation`. It's because it's not schema validation that happens here, that comes later.

Reply:
I understand schema to be a data structure while the errors we get from the failure collector are rather of a logical nature which is not strictly attached to the schema. More generic `Ingress failed validation` would work for both cases IMO if we don't want to have too many messages. Anyway, it's not a critical thing, if you think it's ok, leave it as is 👍